gum/gummemory.c

```diff
@@ -263,54 +288,106 @@ gum_memory_scan (const GumMemoryRange * range,
 }
 
 GumMatchPattern *
-gum_match_pattern_new_from_string (const gchar * match_str)
+gum_match_pattern_new_from_string (const gchar * match_combined_str)
 {
-  GumMatchPattern * pattern;
+  GumMatchPattern * pattern = NULL;
+  gchar ** splitted;
+  const gchar * match_str, * mask_str;
+  gboolean has_mask = FALSE;
   GumMatchToken * token = NULL;
-  const gchar * ch;
+  const gchar * ch, * mh;
+
+  splitted = g_strsplit (match_combined_str, ":", 2);
+  match_str = splitted[0];
+  if (match_str == NULL)
+    goto parse_error;
+
+  mask_str = splitted[1];
+  has_mask = mask_str != NULL;
+  if (has_mask && strlen (mask_str) != strlen (match_str))
+    goto parse_error;
 
   pattern = gum_match_pattern_new ();
 
-  for (ch = match_str; *ch != '\0'; ch++)
+  for (ch = match_str, mh = mask_str;
+       *ch != '\0' && (!has_mask || *mh != '\0');
+       ch++, mh++)
   {
     gint upper, lower;
+    gint mask = 0xff;
     guint8 value;
 
     if (ch[0] == ' ')
       continue;
 
-    if (ch[0] == '?' && ch[1] == '?')
+    if (has_mask)
     {
-      if (token == NULL || token->type != GUM_MATCH_WILDCARD)
-        token = gum_match_pattern_push_token (pattern, GUM_MATCH_WILDCARD);
-      gum_match_token_append (token, 0x42);
-
-      ch++;
-      continue;
+      while (mh[0] == ' ')
+        mh++;
+      if ((upper = g_ascii_xdigit_value (mh[0])) == -1)
+        goto parse_error;
+      if ((lower = g_ascii_xdigit_value (mh[1])) == -1)
+        goto parse_error;
+      mask = (upper << 4) | lower;
     }
 
-    if ((upper = g_ascii_xdigit_value (ch[0])) == -1)
+    if (ch[0] == '?')
+    {
+      upper = 4;
+      mask &= 0x0f;
+    }
+    else if ((upper = g_ascii_xdigit_value (ch[0])) == -1)
+    {
       goto parse_error;
-    if ((lower = g_ascii_xdigit_value (ch[1])) == -1)
+    }
+
+    if (ch[1] == '?')
+    {
+      lower = 2;
+      mask &= 0xf0;
+    }
+    else if ((lower = g_ascii_xdigit_value (ch[1])) == -1)
+    {
       goto parse_error;
+    }
+
     value = (upper << 4) | lower;
 
-    if (token == NULL || token->type != GUM_MATCH_EXACT)
-      token = gum_match_pattern_push_token (pattern, GUM_MATCH_EXACT);
-    gum_match_token_append (token, value);
+    if (mask == 0xff)
+    {
+      if (token == NULL || token->type != GUM_MATCH_EXACT)
+        token = gum_match_pattern_push_token (pattern, GUM_MATCH_EXACT);
+      gum_match_token_append (token, value);
+    }
+    else if (mask == 0x00)
+    {
+      if (token == NULL || token->type != GUM_MATCH_WILDCARD)
+        token = gum_match_pattern_push_token (pattern, GUM_MATCH_WILDCARD);
+      gum_match_token_append (token, 0x42);
+    }
+    else
+    {
+      if (token == NULL || token->type != GUM_MATCH_MASK)
+        token = gum_match_pattern_push_token (pattern, GUM_MATCH_MASK);
+      gum_match_token_append_with_mask (token, value, mask);
+    }
 
     ch++;
+    mh++;
   }
 
   if (!gum_match_pattern_seal (pattern))
     goto parse_error;
 
+  g_strfreev (splitted);
```

Comment:
Nit: a blank line between “unwind” and “return” would improve readability slightly.